### PR TITLE
Implement Gen.choose() for Weighted Distribution

### LINKED ISSUE #974: Weight choose Generators
Add variations on Gen.choose, Gen.enum and so on that allows the weights to be specified.

### PR DESCRIPTION
**Pull request by EarthCitizen**
Port of the algorithm from Haskell QuickCheck for weighted distribution of choices.

Example:
`Gen.choose(1 to 'A', 3 to 'B')` will pick A 25% of the time and B 75% of the time.

One design choice that may or may not be what is wanted is that the function is implemented as:
`Gen.Companion.choose(a: Pair<Int, T>, b: Pair<Int, T>, vararg cs: Pair<Int, T>)`
The reason for this is because it does not make sense to have less than two choices. But, at the same time, one could give the same value for both of the first two arguments.

So, with that said, perhaps it is better just to have only one mandatory parameter even though it would not generate useful results?

Please let me know and I will change it to what ever is wanted.

### CODE REVIEW

**Comment by EarthCitizen** [comment]
I just noticed in the test that I should should have used a linked map instead of a sorted map. Then the ordering makes sense. I will fix that.

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt [hunk 1]
```diff
@@ -490,6 +490,39 @@ fun Gen.Companion.choose(min: Long, max: Long): Gen<Long> {
   }
 }
 
+/**
+ * Returns a stream of values based on weights:
+ *
+ * <pre>Gen.choose(1 to 'A', 2 to 'B')</pre> will generate 'A' 33% of the time
```
Comment (LeoColman) [inline kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt]: I don't think IntelliJ likes `<pre>` or any other tag for Kotlin docs

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt [hunk 2]
```diff
@@ -490,6 +490,39 @@ fun Gen.Companion.choose(min: Long, max: Long): Gen<Long> {
...
+  require(weights.any { it > 0 }) { "At least one weight must be greater than zero"}
+  return object : Gen<T> {
+    // The algorithm for pick is a migration of
+    // the algorithm from Haskell QuickCheck
+    private tailrec fun pick(n: Int, l: Sequence<Pair<Int, T>>): T {
+      val (w, e) = l.first()
+      return if (n <= w) e
+      else pick(n - w, l.drop(1))
+    }
+    override fun constants(): Iterable<T> = emptyList()
```
Comment (LeoColman) [inline kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt:517]: I'm wondering if we should keep the constants as a list of every possible outcome, ignoring weight (but I'm not sure that's even useful)

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt [hunk 3]
```diff
@@ -490,6 +490,39 @@ fun Gen.Companion.choose(min: Long, max: Long): Gen<Long> {
...
+ * weight of zero is given.
+ */
+fun <T : Any> Gen.Companion.choose(a: Pair<Int, T>, b: Pair<Int, T>, vararg cs: Pair<Int, T>): Gen<T> {
+  val allPairs = listOf(a, b) + cs
+  val weights = allPairs.map { it.first }
+  require(weights.all { it >=0 }) { "Negative weights not allowed" }
+  require(weights.any { it > 0 }) { "At least one weight must be greater than zero"}
+  return object : Gen<T> {
+    // The algorithm for pick is a migration of
+    // the algorithm from Haskell QuickCheck
```
Comment (LeoColman) [inline kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt:509]: Perhaps we could add a permalink to that Haskell QuickCheck?

File: kotlintest-assertions/src/jvmTest/kotlin/com/sksamuel/kotlintest/properties/ChooseGenTest.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+         actualRatios[it] shouldBe (expectedRatios[it] plusOrMinus 0.02)
+      }
+    }
+  }
+
+  test("should not accept negative weights") {
+    shouldThrow<IllegalArgumentException> { Gen.choose(-1 to 'A', 1 to 'B') }
+  }
+
+  test("should not accept all zero weights") {
```
Comment (LeoColman) [inline kotlintest-assertions/src/jvmTest/kotlin/com/sksamuel/kotlintest/properties/ChooseGenTest.kt:48]: Perhaps adding a test that checks some zero weights (but not all)?

**Review by LeoColman — APPROVED** [review]
Great pull request!

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt (hunk 1 above)
Comment (EarthCitizen) [reply kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt]: So, remove the pre tags?

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt (hunk 3 above)
Comment (EarthCitizen) [reply kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt:509]: Will add the link.

File: kotlintest-assertions/src/jvmTest/kotlin/com/sksamuel/kotlintest/properties/ChooseGenTest.kt (hunk 4 above)
Comment (EarthCitizen) [reply kotlintest-assertions/src/jvmTest/kotlin/com/sksamuel/kotlintest/properties/ChooseGenTest.kt:48]: Good idea. Will add it.

**Comment by EarthCitizen** [comment]
@Kerooker What is your opinion on the 2 required parameters vs 1 required parameter?

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt (hunk 1 above)
Comment (LeoColman) [reply kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt]: Or try to `ctrl + Q` it and see if it renders correctly

**Comment by LeoColman** [comment]
I mean, calling only `Gen.choose(1 to "b")` doesn't make much sense to me

**Comment by sksamuel** [comment]
Yeah the way you did it with 3 params, last one being varargs makes sense.

**Comment by EarthCitizen** [comment]
I like the 3 params, but the only downside being that someone can do `Gen.choose(1 to 'C', 1 to'C')`.

**Comment by LeoColman** [comment]
> I like the 3 params, but the only downside being that someone can do `Gen.choose(1 to 'C', 2 to'C')`.

I don't think that matters. If you're concerned about it, perhaps we can add a warn that two of the elements are the same

**Comment by ryandanielspmc** [comment]
@Kerooker Probably overkill.

**Comment by sksamuel** [comment]
I don't think it's a big deal either way. The PR looks great.

**Comment by EarthCitizen** [comment]
OK. Will will address @Kerooker 's points and fix something I saw in the tests I submitted.

File: kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt (hunk 2 above)
Comment (EarthCitizen) [reply kotlintest-assertions/src/commonMain/kotlin/io/kotlintest/properties/gens.kt:517]: @Kerooker I am not 100% clear on when the constants need to be present. Can you explain that a bit?

**Comment by EarthCitizen** [comment]
All issues resolved except for the question around whether constants are necessary.

**Comment by sksamuel** [comment]
Constants aren't always needed. There's no edge cases to consider here so I
think its good to merge.
